include/numerics/petsc_matrix.h

```diff
@@ -124,10 +124,12 @@ class PetscMatrix final : public SparseMatrix<T>
    */
   PetscMatrix (PetscMatrix &&) = delete;
   PetscMatrix (const PetscMatrix &) = delete;
-  PetscMatrix & operator= (const PetscMatrix &) = delete;
   PetscMatrix & operator= (PetscMatrix &&) = delete;
   virtual ~PetscMatrix ();
 
+  PetscMatrix & operator= (const PetscMatrix &);
+  virtual SparseMatrix<T> & operator= (const SparseMatrix<T> & v) override;
+
```

Comment:
I know it's not really the main point of the PR, but since we are using virtual assignment operators here, I was wondering if you had [read up on them in detail](https://stackoverflow.com/questions/669818/virtual-assignment-operator-c)? Seems like it might be tricky to get exactly correct.